rust/otap-dataflow/crates/pdata/src/views/otap/logs.rs

```diff
@@ -0,0 +1,2004 @@
+}
+
+/// Extract attribute key from an attribute batch row
+fn get_attribute_key<'a>(cols: &'a OtapAttributeColumns<'a>, row_idx: usize) -> Option<&'a [u8]> {
+    cols.key
+        .as_ref()
+        .and_then(|accessor| accessor.str_at(row_idx))
+        .map(|s| s.as_bytes())
+}
+
+/// Extract attribute value with type information from an attribute batch row
+fn get_attribute_value<'a>(
+    cols: &'a OtapAttributeColumns<'a>,
+    row_idx: usize,
+) -> OtapAnyValueView<'a> {
+    let type_array = match cols.type_id {
+        Some(arr) => arr,
+        None => return OtapAnyValueView::Empty,
+    };
+
+    if !type_array.is_valid(row_idx) {
+        return OtapAnyValueView::Empty;
+    }
+
+    let value_type = match AttributeValueType::try_from(type_array.value(row_idx)) {
+        Ok(t) => t,
+        Err(_) => return OtapAnyValueView::Empty,
+    };
+
+    // Extract the appropriate value based on type using cached accessors
+    match value_type {
+        AttributeValueType::Str => cols
+            .str_val
+            .as_ref()
+            .and_then(|accessor| accessor.str_at(row_idx))
+            .map(|s| OtapAnyValueView::Str(s.as_bytes()))
+            .unwrap_or(OtapAnyValueView::Empty),
+        AttributeValueType::Int => cols
+            .int_val
+            .as_ref()
+            .and_then(|accessor| accessor.value_at(row_idx))
+            .map(OtapAnyValueView::Int)
+            .unwrap_or(OtapAnyValueView::Empty),
+        AttributeValueType::Double => cols
+            .double_val
+            .and_then(|arr| {
+                if arr.is_valid(row_idx) {
+                    Some(OtapAnyValueView::Double(arr.value(row_idx)))
+                } else {
+                    None
+                }
+            })
+            .unwrap_or(OtapAnyValueView::Empty),
+        AttributeValueType::Bool => cols
+            .bool_val
+            .and_then(|arr| {
+                if arr.is_valid(row_idx) {
+                    Some(OtapAnyValueView::Bool(arr.value(row_idx)))
+                } else {
+                    None
+                }
+            })
+            .unwrap_or(OtapAnyValueView::Empty),
+        AttributeValueType::Bytes => cols
+            .bytes_val
+            .as_ref()
+            .and_then(|accessor| accessor.slice_at(row_idx))
+            .map(OtapAnyValueView::Bytes)
+            .unwrap_or(OtapAnyValueView::Empty),
+        _ => OtapAnyValueView::Empty,
+    }
+}
+
+#[allow(dead_code)]
+fn get_schema_url_for_resource<'a>(_batch: &'a RecordBatch, _resource_id: u16) -> Option<Str<'a>> {
+    // TODO: Implement schema_url lookup for resource
+    // Would need to look up in a separate resource table or nested struct
+    None
+}
+
+#[allow(dead_code)]
+fn get_schema_url_for_scope<'a>(_batch: &'a RecordBatch, _scope_id: u16) -> Option<Str<'a>> {
+    // TODO: Implement schema_url lookup for scope
+    // Would need to look up in a separate scope table or nested struct
+    None
+}
+
+#[cfg(test)]
```

Comment:
Did you test scenarios with log batches that contain no resource attributes or fields, or no scope attributes/fields?